Core/Parsing/Structure/IntermediateModel/SupportedMethodSpecifications.cs

```diff
@@ -96,12 +96,27 @@ private static bool HasIndexSelectorParameter (MethodInfo methodInfo, int parame
       if (parameters.Length <= parameterPosition)
         return false;
 
-      var selectorType = parameters[parameterPosition].ParameterType.GetTypeInfo();
-      // Enumerable taks a Func<...> but Querable takes an Expression<Func<...>>
-      if (typeof (Expression).GetTypeInfo().IsAssignableFrom (selectorType))
-        selectorType = selectorType.GenericTypeArguments[0].GetTypeInfo();
+      var selectorType = parameters[parameterPosition].ParameterType.GetTypeInfo().UnwrapEnumerable();
 
       return selectorType.GenericTypeArguments[1] == typeof (int);
     }
+
+    private static TypeInfo UnwrapEnumerable (this TypeInfo typeInfo)
+    {
+      var comparedTypeInfo = typeInfo;
+
+#if !NET_4_0 && !NET_3_5
+      // Workaround for a corner case in .Net Native because
+      // IsAssignableFrom throws for open generics.
+      if (typeInfo.ContainsGenericParameters)
+        comparedTypeInfo = typeInfo.BaseType.GetTypeInfo();
```

Comment:
Good catch! Thought about this, too, when I first saw it but given that it's only used in a very limited context (Enumerable and Queryable) and the type hierarchy employed by the parameter types of Enumerable and Queryable, one hop is enough. Plus, we'll also have unit tests covering 100% of the method lookups, so we're safe against oversights, too.

Now that I re-think this, the only mishap that could happen, were if a new overload would be introduced for the relevant operations and that overload took a parameter at the specific position with more than one open generic hop in it's type hierharchy. Before now, I considered this to be highly unlikely, based on past development of Enumerable and Queryable, but I'm also okay with a loop to find the first non-generic base type (or null). That way, we'd be 100% bullet proof to future changes at this corner of the system.